engine/cdn/cdn_log_store.go

```diff
@@ -14,54 +14,108 @@ import (
 	"github.com/ovh/cds/sdk/log"
 )
 
-func (s *Service) dequeueJobLogs(ctx context.Context) error {
-	defer func() {
-		log.Info(ctx, "cdn: leaving dequeue job logs")
-	}()
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
-			dequeuCtx, cancel := context.WithTimeout(ctx, 15*time.Second)
-			var hm handledMessage
-			if err := s.Cache.DequeueWithContext(dequeuCtx, keyJobLogIncomingQueue, 1*time.Millisecond, &hm); err != nil {
-				cancel()
-				if !strings.Contains(err.Error(), "context deadline exceeded") {
-					log.Error(ctx, "dequeueJobLogs: unable to dequeue job logs queue: %v", err)
+func (s *Service) sendToCDS(ctx context.Context, msgs []handledMessage) error {
+	switch {
+	case msgs[0].Signature.Service != nil:
+		for _, msg := range msgs {
+			// Format line
+			msg.Msg.Full = buildMessage(msg)
+			if msg.Signature.Service != nil {
+				logs := sdk.ServiceLog{
+					ServiceRequirementName: msg.Signature.Service.RequirementName,
+					ServiceRequirementID:   msg.Signature.Service.RequirementID,
+					WorkflowNodeJobRunID:   msg.Signature.JobID,
+					WorkflowNodeRunID:      msg.Signature.NodeRunID,
+					Val:                    msg.Msg.Full,
 				}
-				continue
+				if err := s.Client.QueueServiceLogs(ctx, []sdk.ServiceLog{logs}); err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	default:
+		// Aggregate messages by step
+		hms := make(map[string]handledMessage, len(msgs))
+		for _, msg := range msgs {
+			// Format line
+			msg.Msg.Full = buildMessage(msg)
+
+			k := fmt.Sprintf("%d-%d-%d", msg.Signature.JobID, msg.Signature.NodeRunID, msg.Signature.Worker.StepOrder)
+			// Aggregates lines in a single message
+			if _, ok := hms[k]; ok {
+				full := hms[k].Msg.Full
+				msg.Msg.Full = fmt.Sprintf("%s%s", full, msg.Msg.Full)
+				hms[k] = msg
+			} else {
+				hms[k] = msg
 			}
-			cancel()
+		}
 
-			s.storeLogsWithRetry(ctx, hm)
+		// Send logs to CDS API by step
+		for _, hm := range hms {
+			now := time.Now()
+			l := sdk.Log{
+				JobID:        hm.Signature.JobID,
+				NodeRunID:    hm.Signature.NodeRunID,
+				LastModified: &now,
+				StepOrder:    hm.Signature.Worker.StepOrder,
+				Val:          hm.Msg.Full,
+			}
+			if err := s.Client.QueueSendLogs(ctx, hm.Signature.JobID, l); err != nil {
```

Comment:
Be careful with this, we should not call cds api for each line. But one time with all the lines.

Reply:
Here, we are in a loop that do 1 call by step, so it's OK
Line38->53: aggregation to build 1hm by step
LIne 56 -> 65: 1 call by step

Reply:
ok